src/runmode-af-xdp.c

```diff
@@ -0,0 +1,424 @@
+        SCReturnInt(0);
+    }
+
+    if (StringParseInt32(&aconf->threads, 10, 0, entry_str) < 0) {
+        SCLogError(SC_ERR_INVALID_VALUE,
+                "Threads entry for interface %s contain non-numerical characters - \"%s\"",
+                aconf->iface, entry_str);
+        SCReturnInt(-EINVAL);
+    }
+
+    if (aconf->threads < 0) {
+        SCLogError(SC_ERR_INVALID_VALUE, "Interface %s has a negative number of threads",
+                aconf->iface);
+        SCReturnInt(-ERANGE);
+    }
+
+    if (aconf->threads > nr_queues) {
+        SCLogWarning(SC_WARN_AFXDP_CONF,
+                "Selected threads greater than configured queues, using: %d thread(s)", nr_queues);
+        aconf->threads = nr_queues;
+    }
+
+    SCReturnInt(0);
+}
+
+/**
+ * \brief extract information from config file
+ *
+ * The returned structure will be freed by the thread init function.
+ * This is thus necessary to copy the structure before giving it
+ * to thread or to reparse the file for each thread (and thus have
+ * new structure.
+ *
+ * \return a AFXDPIfaceConfig corresponding to the interface name
+ */
+static void *ParseAFXDPConfig(const char *iface)
+{
+    const char *confstr = NULL;
+    ConfNode *if_root;
+    ConfNode *if_default = NULL;
+    ConfNode *af_xdp_node = NULL;
+    int conf_val = 0;
+    intmax_t conf_val_int = 0;
+    bool boolval = false;
+
+    if (iface == NULL) {
+        return NULL;
+    }
+
+    AFXDPIfaceConfig *aconf = SCCalloc(1, sizeof(*aconf));
+    if (unlikely(aconf == NULL)) {
+        return NULL;
+    }
+
+    memset(aconf, 0, sizeof(*aconf));
+
+    /* default/basic config setup */
+    strlcpy(aconf->iface, iface, sizeof(aconf->iface));
+    aconf->DerefFunc = AFXDPDerefConfig;
+    aconf->threads = 1;
+    aconf->promisc = 1;
+    aconf->busy_poll_time = DEFAULT_BUSY_POLL_TIME;
+    aconf->busy_poll_budget = DEFAULT_BUSY_POLL_BUDGET;
+    aconf->mode = XDP_FLAGS_UPDATE_IF_NOEXIST;
+    aconf->gro_flush_timeout = DEFAULT_GRO_FLUSH_TIMEOUT;
+    aconf->napi_defer_hard_irqs = DEFAULT_NAPI_HARD_IRQS;
+    aconf->mem_alignment = XSK_UMEM__DEFAULT_FLAGS;
+
+    /* Find initial node */
+    af_xdp_node = ConfGetNode("af-xdp");
+    if (af_xdp_node == NULL) {
+        SCLogInfo("unable to find af-xdp config using default values");
+        goto finalize;
+    }
+
+    if_root = ConfFindDeviceConfig(af_xdp_node, iface);
+    if_default = ConfFindDeviceConfig(af_xdp_node, "default");
+
+    if (if_root == NULL && if_default == NULL) {
+        SCLogInfo("unable to find af-xdp config for "
+                  "interface \"%s\" or \"default\", using default values",
+                iface);
+        goto finalize;
+    }
+
+    /* If there is no setting for current interface use default one as main iface */
+    if (if_root == NULL) {
+        if_root = if_default;
+        if_default = NULL;
+    }
+
+    /* Threading */
+    int ret = ConfGetChildValueWithDefault(if_root, if_default, "threads", &confstr) != 1
```

Comment:
I don't like this kind of construct instead of explicit if. It may be just me.

Reply:
That is fine, I can adapt this too. I suppose it makes it more consistent with the overall design anyway.